activation/poetdb.go

```diff
@@ -146,6 +188,16 @@ func (db *PoetDb) GetProofMessage(proofRef types.PoetProofRef) ([]byte, error) {
 
 // Proof returns full proof.
 func (db *PoetDb) Proof(proofRef types.PoetProofRef) (*types.PoetProof, *types.Hash32, error) {
+	// there is no effective rate limiting on gossip layer, and most of the proofs are reused
+	// hence to avoid excessive reads during atx "storm" we do them sequentially
+	// additionally poet proof can be cached during  ValidateAndStore routine
+	db.singleQueryMu.Lock()
+	defer db.singleQueryMu.Unlock()
+	cachedProof, ok := db.poetProofsLru.Get(proofRef)
+	if ok && cachedProof != nil {
+		return &cachedProof.PoetProof, &cachedProof.Statement, nil
+	}
```

Comment:
Isn't allowing only 1 get query at a time too restrictive? Perhaps it could allow only 1 query for given `ref` but allow querying for multiple `ref`s in parallel? It's easy to do with https://pkg.go.dev/golang.org/x/sync/singleflight#Group.Do. Additionally, it could avoid locking on the happy path, possibly reducing contention on the mutex by first trying to get the proof (before locking) like this:

  ```suggestion
	// there is no effective rate limiting on gossip layer, and most of the proofs are reused
	// hence to avoid excessive reads during atx "storm" we do them sequentially
	// additionally poet proof can be cached during  ValidateAndStore routine
	cachedProof, ok := db.poetProofsLru.Get(proofRef)
	if ok && cachedProof != nil {
		return &cachedProof.PoetProof, &cachedProof.Statement, nil
	}
	db.singleQueryMu.Lock()
	defer db.singleQueryMu.Unlock()
	cachedProof, ok := db.poetProofsLru.Get(proofRef)
	if ok && cachedProof != nil {
		return &cachedProof.PoetProof, &cachedProof.Statement, nil
	}
```

What do you think?

Reply:
on the happy path there is always a synchronization in lru cache, so synchronizing before doesn't any overead

as for singleflight i considered doing that, and my opinion was that it is not worth because of 45 proofs for 6_000_000 atxs, but i can change, it won't do anything in terms of measurable improvement but may do if there will be much more proofs

Reply:
>on the happy path there is always a synchronization in lru cache, so synchronizing before doesn't any overead
Makes sense, thanks.

>as for singleflight i considered doing that, and my opinion was that it is not worth because of 45 proofs for 6_000_000 atxs, but i can change, it won't do anything in terms of measurable improvement but may do if there will be much more proofs

OK :+1: